streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -424,6 +426,7 @@
             ValueJoiner<V, V1, R> joiner,
             JoinWindows windows,
             Serde<K> keySerde,
+            Serde<V> thisValSerde,
```

Comment:
Bug. If the stream has been mapped or flatMapped it needs the Serdes passed along otherwise it will fall back to the default which may or may not be correct and you can end up with an exception like below:
[2016-06-13 09:52:15,515] ERROR Streams application error during processing in thread [StreamThread-4]:  (org.apache.kafka.streams.processor.internals.StreamThread:225)
java.lang.ClassCastException: java.lang.Integer cannot be cast to [B